hal/mpu/mbed_mpu_v7m.c

```diff
@@ -0,0 +1,156 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "hal/mpu_api.h"
+#include "cmsis.h"
+
+#if ((__ARM_ARCH_7M__ == 1U) || (__ARM_ARCH_7EM__ == 1U) || (__ARM_ARCH_6M__ == 1U)) && \
+    defined (__MPU_PRESENT) && (__MPU_PRESENT == 1U)
+
+#if !DEVICE_MPU
+#error "Device has v7m MPU but it is not enabled. Add 'MPU' to device_has in targets.json"
+#endif
+
+void mbed_mpu_init()
+{
+    mbed_mpu_enable_ram_xn(false);
+}
+
+void mbed_mpu_free()
+{
+    mbed_mpu_enable_ram_xn(false);
+}
+
+void mbed_mpu_enable_ram_xn(bool enable)
+{
+    // Flush memory writes before configuring the MPU.
+    __DSB();
+
+    const uint32_t regions = (MPU->TYPE & MPU_TYPE_DREGION_Msk) >> MPU_TYPE_DREGION_Pos;
+
+    // Disable the MCU
+    MPU->CTRL = 0;
```

Comment:
Unless we can come up with a better plan for having Ethernet+cache, other code is going to want to set up the MPU, which suggests that this should be much less destructive for MPU state. I guess that would mean:

* Never disabling the MPU
* Never clearing regions we're not setting up
* Having some guidelines about which regions we will use (lowest numbered, presumably) versus those target/drivers can set up (highest numbered, for specific areas?)

Turning off the MPU seems particularly strong - I would have expected the `false` here just to be flipping the XN bit, and any other minimal things needed, not totally change system config.

Reply:
That's very good point.

Reply:
>  Never clearing regions we're not setting up

If this isn't done then bootloader MPU settings may unintentionally persist when an application starts.

> Having some guidelines about which regions we will use (lowest numbered, presumably) versus those target/drivers can set up (highest numbered, for specific areas?)

This may be ok for the ARMv7-M MPU, but this would be problematic for the ARMv8-M MPU which doesn't allow the use of overlapping regions. Access to an address that matches in more than one region generates a precise MemManage fault.

Reply:
> If this isn't done then bootloader MPU settings may unintentionally persist when an application starts.

Yes, but is that necessarily bad? Do we assume the bootloader is helpful or hostile with regard to presetting the MPU? Still, could be resolved by having an early boot clear, then the target init gets a chance to set up, then we don't clear again.

> problematic for the ARMv8-M MPU

Ah, yes, that does greatly change things. I guess then to accomodate ARMv8-M the simplest solution is to have mbed OS core mark the main RAM region determined via linker map as XN WBWA, rather than an all-inclusive address range - no need to be powers of 2 any more - and require that any special regions set up by the target be in a separate named area.

Reply:
> Yes, but is that necessarily bad?

Yes, any implicit dependencies between the bootloader and application should be avoided otherwise you end up with issues like [this](https://github.com/ARMmbed/mbed-os/issues/7382). Applications and bootloaders are typically not updated at the same time, so applications need to be be able to work with all older bootloaders. Other HAL specifications intentionally require the init function to re-initialize the hardware to avoid bugs.

At this point I don't think we want to explicitly require the bootloader to leave the MPU in a specific state so the application can leave it on. For this reason I'm keeping the code which re-initializes the MPU. Let me know if you have concerns with this @kjbracey-arm.

Reply:
To allow vendors to customize things `MPU_CUSTOM` can be defined which stops the common MPU driver from being built. This way vendors can fully customize the MPU for their use case if needed. This should be compatible with the MPU work @jeromecoutant is doing.